### DISCUSSION TITLE
Space shooter

### DISCUSSION THREAD

**Discussion by Clecius** [nCine/nCine#19] [Q&A]
Hello,

In first, compliments for nCine, as far I can see, a good library to create games. =)

I started to create a simple game of space shooter, to try the library.

[Space Shooter](https://i.imgur.com/d1VHVxe.mp4)

I like the library proposal, like a programmer I prefer to PROGRAM my game, not to click and drag on some mumbo jumbo editor that permits some pieces of code.

But I'm having some issues with 'Application Interval' while using VSync activated.
When there's a little more 'Animated Sprites' on the scene they tend to "jump", but this does not occur when VSync is off.
Another issue occurs by the fact that I'm using the 'Application Interval' to count time within the game; to match the weapons shots with the animations, for example. When I turn off VSync the float value of application interval becomes so dirisory that causes the well-known float point error.

[Jumping Sprites](https://i.imgur.com/CIUnIPT.mp4)

Any tips for that?
Currently I'm trying to implement an application interval using a 'long' value with nanoseconds.

There's also a little glitch that appears intermittently on some textures on animations, as can be observed on top of the shield of the main ship on video.

I'm developing on a Debian 11 with GCC 10.2.1.

**Comment by encelo** [comment]
Hello @Clecius, thank you for trying the nCine and well done with your project! 💪🏻 

I see the issue in your Jumping Sprites video and it seems quite weird. I don't remember ever having issues with `Application::interval()` before. 🤔  I see in the video that you have a counter for the total number of frames rendered, are you using `Application::numFrames()` for that?

I would like to understand if the issue is caused by poor performance (but I doubt it with the number of sprites I see on screen) or by some game logic issue.

To better understand if it's a matter of performance you can do a couple things:

1.  Enable the debug overlay
    * Write `config.withDebugOverlay = true;` in the `onPreInit()` callback
2. Enable the Tracy integration
    * Recompile the nCine declaring the `NCINE_WITH_TRACY=ON` CMake variable

If you don't mind, you can also send the source code to me and I will have a look.
 
For a more precise way of measuring the time passed, you can use the `Timestamp` class, which internally uses a `uint64_t` counter of ticks from the Linux `CLOCK_MONOTONIC`.

For the glitch on the ship shield, I would say it is caused by missing padding pixels between the elements of your spritesheet. You are using a texture atlas for the shield, right?

Don't hesitate to ask if you need more help and consider joining our [Discord](https://discord.gg/495ab6Y) server. 👍🏻

**Reply by Clecius** [reply]
Hello  @encelo , sorry for the delay, too many projects. 😕
I'm using just my spare time on this project.

I fixed some issues on the battle mechanics, but the Jumping Sprites still persist.
I enabled the Debug Overlay and recompiled nCine with `NCINE_WITH_TRACY=ON`, but cannot find anything abnormal.
I'll send you the source code for your analysis.

Thanks for the tip with `Timestamp`, that really saved me some headache. 😁
The C++ `Chrono` library has some problems between hardwares. 😒

About the glitching sprites; yes, I'm using texture atlas, to be specific I'm using the **TexturePacker**.
Adding the padding pixel solved for small sprites, but not for background (big images); I will try this one without texture atlas.


I have another question.
I'm trying to build nCine using Nuklear, by adding `NCINE_WITH_NUKLEAR=ON` on CMake, but I'm getting a compile error on NuklearDrawing.cpp.

```
/develop/lib_3rd/cpp/source/nCine/nCine/src/graphics/NuklearDrawing.cpp: In member function ‘bool ncine::NuklearDrawing::bakeFonts()’:
/develop/lib_3rd/cpp/source/nCine/nCine/src/graphics/NuklearDrawing.cpp:102:10: error: ‘struct nk_convert_config’ has no member named ‘null’
102 |  config_.null = null_;
    |          ^~~~
```

Am I missing something?

**Reply by encelo** [reply]
Hello @Clecius, yes please, send me the source code whenever you want, I'm curious about the jumping issue. 🤔 

I'm happy to hear that the `Timestamp` class is working fine for you. 🥳 

Strange to see that the background is still presenting glitches. When I will try the game I will have a look into this as well.

You are right about Nuklear, I haven't tested it for some time and it seems they have renamed a field in the configuration structure (https://github.com/Immediate-Mode-UI/Nuklear/commit/ec4acc7cf4e6e53d604d17e6cab5019233a221cc). Just rename the `null` field to `tex_null` and it should compile again. 👌🏻
